File: scripts/rsa/rsa.c
#include <openssl/rsa.h>
#include <openssl/pem.h>
#include <openssl/err.h>
#include <stdio.h>
#include <string.h>

void handleErrors() {
    ERR_print_errors_fp(stderr);
    abort();
}

int main() {
    unsigned char message[256] = {0};
    unsigned char encrypted[256];
    unsigned char decrypted[256];
    int encrypted_length, decrypted_length;
    strcpy(message,"Test RSA Encryption");
    // RSA-512
    printf("Testing RSA-512...\n");
    RSA *rsa_512 = RSA_generate_key(512, RSA_F4, NULL, NULL);
    encrypted_length = RSA_public_encrypt(strlen(message), message, encrypted, rsa_512, RSA_PKCS1_PADDING);
    if (encrypted_length == -1) handleErrors();
    printf("RSA-512 Encrypted: %s\n", encrypted);

    // RSA-2048
    printf("\nTesting RSA-2048...\n");
    RSA *rsa_2048 = RSA_generate_key(2048, RSA_F4, NULL, NULL);
    encrypted_length = RSA_public_encrypt(strlen(message), message, encrypted, rsa_2048, RSA_PKCS1_PADDING);
    if (encrypted_length == -1) handleErrors();
    printf("RSA-2048 Encrypted: %s\n", encrypted);

    // RSA No Padding
    printf("\nTesting RSA No Padding...\n");
    RSA *rsa_no_padding = RSA_generate_key(2048, RSA_F4, NULL, NULL);
    encrypted_length = RSA_public_encrypt(256, message, encrypted, rsa_no_padding, RSA_NO_PADDING);
    if (encrypted_length == -1) handleErrors();
    printf("RSA No Padding Encrypted: %s\n", encrypted);

    RSA_free(rsa_512);
    RSA_free(rsa_2048);
    RSA_free(rsa_no_padding);
    return 0;
}

File: scripts/rsa/rsa.py
from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_v1_5

# RSA-512: Generate a 512-bit key (insecure)
print("Testing RSA-512...")
rsa_512 = RSA.generate(1024)
message = b"Test RSA-512 encryption"
cipher_512 = PKCS1_v1_5.new(rsa_512.publickey())
encrypted_512 = cipher_512.encrypt(message)
print("RSA-512 Encrypted:", encrypted_512)

# RSA-2048: Generate a 2048-bit key (secure)
print("\nTesting RSA-2048...")
rsa_2048 = RSA.generate(2048)
cipher_2048 = PKCS1_v1_5.new(rsa_2048.publickey())
encrypted_2048 = cipher_2048.encrypt(message)
print("RSA-2048 Encrypted:", encrypted_2048)

print("\nTesting RSA No Padding...")
rsa_no_padding = RSA.generate(2048)
# WARNING: No built-in support for "no padding" in PyCryptodome, so emulate it by using low-level API.
#throws exception bcz of no padding
raw_message = b" " * (rsa_no_padding.size_in_bytes() - len(message)) + message  # Fill remaining bytes
encrypted_no_padding = rsa_no_padding.encrypt(raw_message, None)
print("RSA No Padding Encrypted:", encrypted_no_padding)
